tls/s2n_security_policies.c

```diff
@@ -891,6 +900,30 @@ int s2n_config_set_cipher_preferences(struct s2n_config *config, const char *ver
     return 0;
 }
 
+static bool s2n_cipher_suite_uses_chacha20_alg(struct s2n_cipher_suite *cipher_suite) {
+    if (!cipher_suite) {
+        return false;
+    }
+
+    /* If s2n_chacha20_poly1305 is available, then we can directly check the record_alg */
+    if (s2n_chacha20_poly1305.is_available()) {
+        return cipher_suite->record_alg && cipher_suite->record_alg->cipher == &s2n_chacha20_poly1305;
+    } 
+
+    /* If s2n_chacha20_poly1305 is not available then cipher_suite->record_alg is null, so that field can't be
+     * used to check if the ciphersuite is ChaCha20. However we can iterate over cipher_suite->all_record_algs,
+     * as a ChaCha20 ciphersuite will always have a ChaCha20 record algorithm in that list.
+     */
+    for (size_t i = 0; i < cipher_suite->num_record_algs; i++) {
+        const struct s2n_record_algorithm* rec_alg = cipher_suite->all_record_algs[i];
+        if (rec_alg && rec_alg->cipher == &s2n_chacha20_poly1305) {
+            return true;
+        }
+    }
```

Comment:
I still think this is an unnecessary complication. We should just skip the "invalid" security policy check if chacha20 isn't supported.

An "invalid" security policy doesn't break any logic. So it's technically not even really invalid, it's just likely the author made a mistake like forgetting to remove the flag when they removed chacah20 cipher suites. But ignoring that, our CI will catch any "invalid" security policies before they make it into the codebase. And before the CI, developers will probably catch any "invalid" security policies locally, because the most common defaults (Openssl-1.1.1 and awslc) both support chacha20.

Further, does this even need to be an init check? Should we just check all hardcoded security policies in a unit test and avoid needing to do this every time the library is initialized?

Reply:
As discussed offline; moved the chacha20 configuration check to the unit tests. This check will only happen when chacha20 is available in the libcrypto